Updates on VF2++ (#80)
scientific-python/blog.scientific-python.org: content/posts/networkx/node-ordering-Ti-updating/index.md

$$O(2N_mD + 2N) = O(N_mD + N)$$

in which we have excluded the lookup times in $T_i$, $mapping$ and $reverse\_mapping$ as they are all $O(1)$. Our
incremental method works like this:

```python
def update_Tinout(
    G1, G2, T1, T2, T1_out, T2_out, new_node1, new_node2, mapping, reverse_mapping
):
    # This function should be called right after the feasibility is established and node1 is mapped to node2.
    uncovered_neighbors_G1 = {nbr for nbr in G1[new_node1] if nbr not in mapping}
    uncovered_neighbors_G2 = {
        nbr for nbr in G2[new_node2] if nbr not in reverse_mapping
    }

    # Add the uncovered neighbors of node1 and node2 in T1 and T2 respectively
    T1.discard(new_node1)
    T2.discard(new_node2)
    T1 = T1.union(uncovered_neighbors_G1)
    T2 = T2.union(uncovered_neighbors_G2)

    # todo: maybe check this twice just to make sure
    T1_out.discard(new_node1)
    T2_out.discard(new_node2)
    T1_out = T1_out - uncovered_neighbors_G1
    T2_out = T2_out - uncovered_neighbors_G2

    return T1, T2, T1_out, T2_out
```

which based on the previous notation, is:

$$O(2D + 2(D + M_{T_1}) + 2D) = O(D + M_{T_1})$$

> where $M_{T_1}$ is the expected (average) number of elements in $T_1$. Certainly, the complexity is much better in this
> case, as $D$ and $M_{T_1}$ are significantly smaller than $N_mD$ and $N$.

Reviewer: The conclusion should stand out
```suggestion
where $M_{T_1}$ is the expected (average) number of elements in $T_1$.

Certainly, the complexity is much better in this
case, as $D$ and $M_{T_1}$ are significantly smaller than $N_mD$ and $N$.
```

observations to implement an incremental computation of $T_i$ and $\tilde{T_i}$ and make VF2++ more efficient.

- $T_i$ is empty in the beggining, since there are no mapped nodes ($m=\varnothing$) and therefore no neighbors of
  mapped nodes.
- $\tilde{T_i}$ initially contains all the nodes from graph $G_i, i=1,2$ which can be realized directly from the
  notation if we consider both $m$ and $T_1$ empty sets.
- Every step of the algorithm either adds one node $u$ to the mapping or pops one from it.

We can conclude that in every step, $T_i$ and $\tilde{T_i}$ can be incrementally updated. This method avoids a ton of
redundant operations and results in significant performance improvement.

![Performance comparison between brute force Ti computing and incremental updating.](acceleration.png)

The above graph shows the difference in performance between using the exhaustive brute force and incrementally updating
$T_i$ and $\tilde{T_i}$. The graph used to obtain these measurements was a regular
[GNP Graph](https://en.wikipedia.org/wiki/Erd%C5%91s%E2%80%93R%C3%A9nyi_model) with a probability for an edge equal to

> $0.7$. It is directly observed that the increase in time when using the brute force method, follows an exponential rate,
> while in our method, it's not significantly affected by incrementing the number of nodes. Let's talk time complexity.

Reviewer 2: I think this statement doesn't *quite* match what the performance plot shows. For instance, both methods *appear* to have an exponential increase in run-time vs. number of nodes, it's just that the incremental update method increases much less rapidly. Perhaps something like:

```suggestion
$0.7$. It can clearly be seen that execution time of the brute force
method increases much more rapidly with the number of nodes/edges than
the incremental update method, as expected.
```

Author: Sure, you are right!

incremental method works like this:

```python
def update_Tinout(
    G1, G2, T1, T2, T1_out, T2_out, new_node1, new_node2, mapping, reverse_mapping
):
    # This function should be called right after the feasibility is established and node1 is mapped to node2.
    uncovered_neighbors_G1 = {nbr for nbr in G1[new_node1] if nbr not in mapping}
    uncovered_neighbors_G2 = {
        nbr for nbr in G2[new_node2] if nbr not in reverse_mapping
    }

    # Add the uncovered neighbors of node1 and node2 in T1 and T2 respectively
    T1.discard(new_node1)
    T2.discard(new_node2)
    T1 = T1.union(uncovered_neighbors_G1)
    T2 = T2.union(uncovered_neighbors_G2)

    # todo: maybe check this twice just to make sure
    T1_out.discard(new_node1)
    T2_out.discard(new_node2)
    T1_out = T1_out - uncovered_neighbors_G1
    T2_out = T2_out - uncovered_neighbors_G2

    return T1, T2, T1_out, T2_out
```

which based on the previous notation, is:

$$O(2D + 2(D + M_{T_1}) + 2D) = O(D + M_{T_1})$$

where $M_{T_1}$ is the expected (average) number of elements in $T_1$.

Certainly, the complexity is much better in this

> case, as $D$ and $M_{T_1}$ are significantly smaller than $N_mD$ and $N$.

Reviewer 2: Since you're already working on the next post, it might be nice to wrap this one up in a sentence or two and give readers a teaser about what's coming next! e.g. `In this post we've investigated how node ordering is implemented for the VF2++ algorithm; next we'll investigate another important compenent: cuts and consistency checks <or whatever your next post will be about)`

---
title: "Updates on VF2++"
date: 2022-07-06
draft: false
description: "
Summary of the progress on VF2++
"
tags: ["gsoc", "networkx"]
displayInList: true
author: ["Konstantinos Petridis"]

---

> This post includes all the major updates since the last post about VF2++. Each section is dedicated to a different

Reviewer: Add a link to the previous post. You can have relative links `[here]({{< relref "../GSoC-2022" >}}`.

I would also update the first post with a table of content at the beginning. We can have the same table in all posts. Something similar is found in other blogs when there are follow up.

>      - $V_d=\[$nodes of the $d^{th}$ level$\]$
>      - $\bar{V_1} \setminus V_d$
>      - $ProcessLevel(V_d)$
> 4. Output $M$: the matching order of the nodes.

> ---
>
> **Process Level**
>
> ---
>
> 1. **while** $V_d$ not empty **do**
>    - $S=\[$nodes from $V_d$ with the most neighbors in M$\]$
>    - $maxNodes=argmax_{degree}\(S\)$
>    - $rarestNode=\[$node from $maxNodes$ with the rarest label$\]$
>    - $V_d \setminus m$
>    - Append m to M

## $T_i$ and $\tilde{T_i}$

According to the VF2++ paper notation:

$$T_1=\(u\in V_1 \setminus m: \exists \tilde{u} \in m: (u,\tilde{u}\in E_1)\)$$

where $V_1$ and $E_1$ contain all the nodes and edges of the first graph respectively, and $m$ is a dictionary, mapping
every node of the first graph to a node of the second graph. Now if we interpet the above equation, we conclude that
$T_1$ contains uncovered neighbors of covered nodes. In simple terms, it includes all the nodes that do not belong in
the mapping $m$ yet, but are neighbors of nodes that are in the mapping. In addition,

$$\tilde{T_1}=\(V_1 \setminus m \setminus T_1\)$$

The following figure is meant to provide some visual explanation of what exactly $T_i$ is.

> <center><img src="Ti.png" alt="Illustration of $T_i$"/></center>

Reviewer: Is there a reason you did not use normal markdown here?
```suggestion
![Illustration of $T_i$.](Ti.png)
```

Author: nope, i just took it as is from another post. I am not an md expert unfortunately :)

edges are just the neighbors of the covered (mapped) nodes. Here, $T_1$ contains the red nodes (4,5,6) which are
neighbors of the covered nodes 1,2,3, and $T_2$ contains the grey ones (D,E,F). None of the nodes depicted would be
included in $\tilde{T_1}$ or $\tilde{T_2}$. The latters would contain all the remaining nodes from the two graphs.

Regarding the computation of these sets, it's not practical to use the brute force method and iterate over all nodes in
every step of the algorithm to find the desired nodes and compute $T_i$ and $\tilde{T_i}$. We use the following
observations to implement an incremental computation of $T_i$ and $\tilde{T_i}$ and make VF2++ more efficient.

- $T_i$ is empty in the beggining, since there are no mapped nodes ($m=\varnothing$) and therefore no neighbors of
  mapped nodes.
- $\tilde{T_i}$ initially contains all the nodes from graph $G_i, i=1,2$ which can be realized directly from the
  notation if we consider both $m$ and $T_1$ empty sets.
- Every step of the algorithm either adds one node $u$ to the mapping or pops one from it.

We can conclude that in every step, $T_i$ and $\tilde{T_i}$ can be incrementally updated. This method avoids a ton of

> redundant operations and results in significant performance improvement.

Reviewer: Thank you for the update. If you can regenerate the figure, we need to make it more accessible by using different linestyles and adding a legend.
